Worker shutdown notification middleware

rpkilby opened the pull request:
Adds a middleware that notifies actors when their worker process has been signaled for termination. Like the `TimeLimit` (and proposed task cancellation) middleware, an exception is raised in the actor so that the code can properly clean up after itself. Actors opt into the behavior by setting the `notify_shutdown` option to True. If the option is set, the `Shutdown` exception is raised in the actor whenever its worker process is sent a sigterm. If the option is not set, then the exception is not raised.

This PR consists of two chunks:

1) Factor out the exception raising behavior of `TimeLimit` into a new `ThreadExceptingMiddleware` 
    - (might also be useful for the proposed `RemoteControlMiddleware` from #37).
2) Add the `ShutdownNotifications` middleware, which inherits this new middleware base class. 


At this point the PR is mostly ready, but I wanted to make sure this was worth pursing before cleaning it up.

**Possible todos:**
- Review naming. eg, is `ThreadExceptingMiddleware` ideal?
- more tests
- add docs
- add example

Bogdanp commented:
Took a quick peek at this and it looks great! I'll do a more in-depth review tonight.

Bogdanp commented on dramatiq/middleware/shutdown.py:
```diff
@@ -0,0 +1,69 @@
+        thread_id = threading.get_ident()
+
+        if thread_id in self.notifications:
+            self.notifications.remove(thread_id)
```
You should add `after_skip_message = after_process_message` underneath this.

rpkilby replied:
Should this also apply to the `TimeLimit` middleware?

Bogdanp replied:
Yes, it should. Looks like I missed it when I added `SkipMessage`.

Bogdanp commented on dramatiq/middleware/threading.py:
```diff
@@ -0,0 +1,67 @@
+from .middleware import Middleware
+
+
+class ThreadExceptingMiddleware(Middleware):
```
I think it might be better not to rely on inheritance here. Just extract `raise_thread_exception` as a normal function in this module that the middleware can use. We can leave out the platform warning or we can have each of the middleware implement it individually, I don't see much value in DRYing that up, especially if it means having to follow the inheritance tree up in order to figure out where it's firing from.

rpkilby commented:
Hi @Bogdanp. I've updated the code and rebased/reworked the commit history. The code changes should be ready for review, although I'm still working on updating the docs and adding an example.

- Thread raising reworked from base class to function
- Added base `Interrupt` exception for `TimeLimitExceeded` and `Shutdown`. In general, this seems like a good term to describe the behavior. These middleware are asynchronously raising an exception on worker threads in order to interrupt the actor's execution. 
- Added `tests/middleware` to house the threading and shutdown tests. I've moved the prometheus tests into there as well. Does that work, or do you prefer having them all in a flat tests directory?
- Added `ShutdownNotifications to the `default_middleware`, and moved `default_middleware` into the `middleware` package.

rpkilby commented on dramatiq/middleware/shutdown.py:
```diff
@@ -0,0 +1,83 @@
+    def before_worker_shutdown(self, broker, worker):
+        self.logger.debug("Sending shutdown notification to worker threads...")
+        for thread_id in self.notifications:
+            self.logger.info("Worker shutdown notification. Raising exception in worker thread %r.", thread_id)
```
Is it relevant to include the `thread_id` here, since the worker's log format includes the thread name?

Bogdanp replied:
I believe so, because the thread logging the message may be different from the thread executing the task.

rpkilby commented:
Alright, added docs and an example. Everything should be ready for review now.

Bogdanp commented:
Thanks again! I merged this with rebase here: https://github.com/Bogdanp/dramatiq/compare/d323bb027be6187dafea793e3010a40a0fe4eebf...5c21bb7
